test/jdk/java/awt/ScrollPane/ComponentScrollTest/ComponentScrollTest.java

```diff
@@ -0,0 +1,99 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+
+/*
+  @test
+  @bug 4342129
+  @summary Unable to scroll in scrollpane for canvas
+  @key headful
+*/
+
+import java.awt.Component;
+import java.awt.Dimension;
+import java.awt.EventQueue;
+import java.awt.Frame;
+import java.awt.Graphics;
+import java.awt.ScrollPane;
+
+import java.awt.event.AdjustmentEvent;
+import java.awt.event.AdjustmentListener;
+
+import java.lang.reflect.InvocationTargetException;
+
+public class ComponentScrollTest implements AdjustmentListener {
+    public ScrollPane scrollpane;
+    public Frame frame;
+    public int count = 0;
+    public static void main(String[] args) throws InterruptedException, InvocationTargetException {
+        ComponentScrollTest cst = new ComponentScrollTest();
+        cst.init();
+        cst.start();
+    }
+
+    public void init() throws InterruptedException, InvocationTargetException {
+        EventQueue.invokeAndWait(() -> {
+                scrollpane = new ScrollPane();
+                frame = new Frame("Component Scroll Test");
+                scrollpane.add(new Component() {
+                    public Dimension getPreferredSize() {
+                        return new Dimension(500, 500);
+                    }
+                    public void paint(Graphics g) {
+                        g.drawLine(0, 0, 500, 500);
+                    }
+                });
+                frame.add(scrollpane);
+        });
+        scrollpane.getVAdjustable().addAdjustmentListener(this);
+    }
+
+    public void start() throws InterruptedException, InvocationTargetException {
+        try {
+            EventQueue.invokeAndWait(() -> {
+                frame.pack();
+                frame.setVisible(true);
+                scrollpane.getVAdjustable().setValue(20);
+                try {
+                    Thread.sleep(5000);
+                } catch (InterruptedException ie) {
+                }
+                System.out.println("Count = " + count);
+                if (count > 50) {
+                    throw new RuntimeException();
+                }
+            });
+        } finally {
+            EventQueue.invokeLater(() -> {
```

Comment:
Why the invokeLater is used here?

Reply:
I preferred asynchronous EDT would be better since its finally block. Is there any particular reason that we should use `invokeAndWait` than `invokeLater`?

Reply:
> I preferred asynchronous EDT would be better since its finally block. Is there any particular reason that we should use `invokeAndWait` than `invokeLater`?

You want to **wait** until the frame is disposed of before exiting the main thread otherwise it doesn't make sense.

Reply:
yes, I just noticed this in a different proposed change and I agree wait makes more sense.

Reply:
Ok, didn't think in that sense, I'll better use invokeAndWait.....